### PR TITLE
Update delegation

### PR DESCRIPTION
**Pull request by edytapawlak**
Updates delegation to work as described [here](https://github.com/decentralized-identity/keri/issues/146)

It also includes some changes not directly related to delegation:
* removing `log` module because it's depreciated,
* separation of `signed_event_message` submodule in `event_message` mod

### CODE REVIEW

File: src/derivation/attached_signature_code.rs [hunk 1]
```diff
@@ -1,6 +1,6 @@
 use super::{self_signing::SelfSigning, DerivationCode};
-use crate::error::Error;
-use base64::{decode_config, encode_config};
+use crate::{error::Error, event_message::payload_size::num_to_b64};
+use base64::{decode_config};
```
Comment (35359595) [inline src/derivation/attached_signature_code.rs:3]: braces here are no longer required

File: src/event/event_data/delegated/mod.rs [hunk 2]
```diff
@@ -14,17 +14,14 @@ pub struct DelegatedInceptionEvent {
...
 
-    #[serde(rename = "da")]
-    pub seal: LocationSeal,
+    #[serde(rename = "di")]
+    pub delegator: IdentifierPrefix,
 }
 
 #[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
 pub struct DelegatedRotationEvent {
     #[serde(flatten)]
```
Comment (35359595) [inline src/event/event_data/delegated/mod.rs:23]: if there is no difference now between Delegation and Rotation events why keep both? can we use RotationEvent struct wherever needed?

File: src/event_message/attachement.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+use crate::{
+    error::Error,
+    prefix::{Prefix, SelfAddressingPrefix},
+};
+
+use super::{parse::counter, payload_size::PayloadType};
+
+#[derive(Debug, Clone, Deserialize, PartialEq)]
+pub enum Attachement {
```
Comment (35359595) [inline src/event_message/attachement.rs:14]: typo in name? Should be `Attachment`?

File: src/event_message/attachement.rs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            "-G" => {
+                let (rest, counter) = counter(s.as_bytes())
+                    .map_err(|_e| Error::SemanticError("Can't parse counter".into()))?;
+                if rest.is_empty() {
+                    Ok(counter)
+                } else {
+                    Err(Error::SemanticError("Can't parse counter".into()))
+                }
+            }
+            _ => todo!(),
```
Comment (35359595) [inline src/event_message/attachement.rs:53]: should we wait for this to be implemented before merging?

File: src/event_message/attachement.rs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct SorceSeal {
```
Comment (35359595) [inline src/event_message/attachement.rs:59]: typo in name? Should be `SourceSeal`?

File: src/event_message/attachement.rs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            .join("")
+            .as_bytes()
+            .to_vec())
+    }
+}
+
+fn pack_sn(sn: u64) -> String {
+    let payload_type = PayloadType::OA;
+    let sn_raw: Vec<u8> = sn.to_be_bytes().into();
+    // Calculate how many zeros are missing to achive expected base64 string
```
Comment (35359595) [inline src/event_message/attachement.rs:79]: typo? should be `to achieve expected`?

File: src/event_message/attachement.rs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    let sn_vec: Vec<u8> = std::iter::repeat(0).take(missing_zeros).chain(sn_raw).collect();
+    [
+        payload_type.to_string(),
+        base64::encode_config(sn_vec, URL_SAFE_NO_PAD),
+    ]
+    .join("")
+}
+
+#[test]
+fn test_parse_attachement() -> Result<(), Error> {
```
Comment (35359595) [inline src/event_message/attachement.rs:91]: also typo `attachment`? =)

File: src/event_message/event_msg_builder.rs [hunk 8]
```diff
@@ -107,9 +98,9 @@ impl EventMsgBuilder {
         EventMsgBuilder { ..self }
     }
 
-    pub fn with_delegating_seal(self, seal: LocationSeal) -> Self {
+    pub fn with_delegator(self, delegator: IdentifierPrefix) -> Self {
```
Comment (35359595) [inline src/event_message/event_msg_builder.rs:101]: should this take ownership on `delegator` or ref + clone? how's calling code uses it?

Comment (35359595) [reply src/event_message/event_msg_builder.rs:101]: Maybe this also applyes to below `SignatureThreshold` too?

File: src/event_message/parse.rs [hunk 9]
```diff
@@ -1,20 +1,5 @@
...
-use crate::{
-    derivation::attached_signature_code::b64_to_num,
-    event::{event_data::EventData, sections::seal::EventSeal},
-    prefix::{
-        parse::{attached_signature, basic_prefix, event_seal, self_signing_prefix},
-        BasicPrefix, SelfSigningPrefix,
-    },
-    state::IdentifierState,
-};
+use super::{AttachedSignaturePrefix, EventMessage, SignedEventMessage, attachement::{SorceSeal, Attachement}, payload_size::PayloadType, signed_event_message::{SignedNontransferableReceipt, SignedTransferableReceipt}};
```
Comment (35359595) [inline src/event_message/parse.rs:1]: can we agree on import formatting?
Single line vs one import per line - it's kinda mess in every module.
Auto add via IDE will flatten everything into single line, so I guess we can go with that?

File: src/event_message/parse.rs [hunk 10]
```diff
@@ -38,12 +23,13 @@ pub struct DeserializedEvent<'a> {
 pub struct DeserializedSignedEvent<'a> {
     pub deserialized_event: DeserializedEvent<'a>,
     pub signatures: Vec<AttachedSignaturePrefix>,
+    pub attachement: Option<Attachement>,
```
Comment (35359595) [inline src/event_message/parse.rs:26]: typo in prop name? should be `attachment`?

File: src/event_message/parse.rs [hunk 11]
```diff
@@ -38,12 +23,13 @@ pub struct DeserializedEvent<'a> {
...
 }
 
 // FIXME: detect payload type
 impl From<DeserializedSignedEvent<'_>> for SignedEventMessage {
     fn from(de: DeserializedSignedEvent) -> SignedEventMessage {
-        SignedEventMessage::new(&de.deserialized_event.event_message, PayloadType::MA, de.signatures)
+        SignedEventMessage::new(&de.deserialized_event.event_message, PayloadType::MA, de.signatures, de.attachement)
```
Comment (35359595) [inline src/event_message/parse.rs:32]: same typo?

File: src/event_message/payload_size.rs [hunk 12]
```diff
@@ -77,11 +80,11 @@ pub enum PayloadType {
     // TODO: Indexed signatures
 }
 
-#[cfg(feature = "async")]
+// #[cfg(feature = "async")]
```
Comment (35359595) [inline src/event_message/payload_size.rs:83]: delete this line if used always =)

File: src/event_message/signed_event_message.rs [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+// KERI serializer should be used to serialize this
+#[derive(Debug, Clone, Deserialize)]
+pub struct SignedEventMessage {
+    pub event_message: EventMessage,
+    #[serde(skip_serializing)]
+    pub payload_type: PayloadType, 
+    #[serde(skip_serializing)]
+    pub signatures: Vec<AttachedSignaturePrefix>,
+    #[serde(skip_serializing)]
+    pub attachement: Option<Attachement>,
```
Comment (35359595) [inline src/event_message/signed_event_message.rs:19]: typo again?

File: src/event_message/signed_event_message.rs [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+                false => Ordering::Less,
+            },
+        }
+    }
+}
+
+impl Eq for TimestampedSignedEventMessage {}
+
+impl SignedEventMessage {
+    pub fn new(message: &EventMessage, payload_type: PayloadType, sigs: Vec<AttachedSignaturePrefix>, attachement: Option<Attachement>) -> Self {
```
Comment (35359595) [inline src/event_message/signed_event_message.rs:122]: typo in `attachment`?

**Review by 35359595 — CHANGES_REQUESTED** [review]
Everything else looks good =)

File: src/event_message/event_msg_builder.rs (hunk 8 above)
Comment (edytapawlak) [reply src/event_message/event_msg_builder.rs]: I've updated it for both. Now delegator prefix and signature threshold are passed by ref :)

File: src/event_message/parse.rs (hunk 9 above)
Comment (edytapawlak) [reply src/event_message/parse.rs]: This can be solved by using `cargo fmt`. Should I reformat it here or it's better to have separate PR for that?

**Review by 35359595 — APPROVED** [review]
All good. Can merge.

**Pull request merged**